lib/chef/resource/locale.rb

```diff
@@ -36,56 +36,41 @@ class Locale < Chef::Resource
       action :update do
         description "Update the system's locale."
 
-        if node["init_package"] == "systemd"
-          # on systemd settings LC_ALL is (correctly) reserved only for testing and cannot be set globally
-          execute "localectl set-locale LANG=#{new_resource.lang}" do
-            # RHEL uses /etc/locale.conf
-            not_if { up_to_date?("/etc/locale.conf", new_resource.lang) } if ::File.exist?("/etc/locale.conf")
-            # Ubuntu 16.04 still uses /etc/default/locale
-            not_if { up_to_date?("/etc/default/locale", new_resource.lang) } if ::File.exist?("/etc/default/locale")
-          end
+        if ::File.exist?("/etc/default/locale")
+          locale_file = "/etc/default/locale"
         elsif ::File.exist?("/etc/sysconfig/i18n")
-          locale_file_path = "/etc/sysconfig/i18n"
-
-          updated = up_to_date?(locale_file_path, new_resource.lang, new_resource.lc_all)
+          locale_file = "/etc/sysconfig/i18n"
+        elsif ::File.exist?("/etc/environment")
+          locale_file = "/etc/environment"
+        else
+          raise "#{node["platform"]} platform not supported by the chef locale resource."
+        end
 
-          file locale_file_path do
-            content lazy {
-              locale = IO.read(locale_file_path)
-              variables = Hash[locale.lines.map { |line| line.strip.split("=") }]
-              variables["LANG"] = new_resource.lang
-              variables["LC_ALL"] =
-                variables.map { |pairs| pairs.join("=") }.join("\n") + "\n"
-            }
-            not_if { updated }
-          end
+        contents = IO.readlines(locale_file)
+        env_val = contents.map { |t| t.split("=") if t.include?("=") }.compact.to_h
```

Comment:
["k=1", "", "k2=2"].map { |t| t.split("=") if t.include?("=") }.to_h

3: from /home/marc/.rubies/ruby-2.5.3/bin/irb:11:in `<main>'
2: from (irb):3
1: from (irb):3:in `to_h'
TypeError (wrong element type NilClass at 1 (expected array))
irb(main):004:0> 
```

Reply:
This also deletes comments, which we probably shouldn't do:

```
$ cat /etc/default/locale 
#  File generated by update-locale
LANG="en_US.UTF-8"

$ irb
2.5.1 :001 > locale = IO.readlines("/etc/default/locale")
 => ["#  File generated by update-locale\n", "LANG=\"en_US.UTF-8\"\n"] 
2.5.1 :002 > locale.map { |t| t.split("=") if t.include?("=") }.compact.to_h
 => {"LANG"=>"\"en_US.UTF-8\"\n"} 
2.5.1 :003 > 
```

> [Ruby] Hashes enumerate their values in the order that the corresponding keys were inserted.

We would have to read the comments into our hash in the same loop as the `map` though to maintain the order.

Reply:
Tried and updated the logic. Handled comments, and even any invalid lines, they will remain untouched. Added the specs for almost around all the cases.
Could you please take a look and let me know in case of any issues.